include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,436 @@
+  /// in which case its shape would be 2-D (values_count, 2).
+  ///
+  /// Values shape is supplied at construction time and its Size() must match values_count.
+  /// </summary>
+  /// <param name="values_count"></param>
+  /// <param name="values_data"></param>
+  /// <param name="indices"></param>
+  /// <returns></returns>
+  Status MakeCooData(const IDataTransfer& data_transfer, const OrtMemoryInfo& data_location,
+                     size_t values_count, const void* values_data, gsl::span<const int64_t> indices);
+
+  /// <summary>
+  /// Gives mutable access to Coo buffers so they can be populated
+  /// </summary>
+  class CooMutator {
+   public:
+    CooMutator(Tensor& values, Tensor& indices) noexcept : values_(values), indices_(indices) {}
+    Tensor& Values() noexcept { return values_; }
+    Tensor& Indices() noexcept { return indices_; }
+
+   private:
+    std::reference_wrapper<Tensor> values_;
+    std::reference_wrapper<Tensor> indices_;
+  };
+
+  /// <summary>
+  /// Allocates memory for values and index and returns a mutator so
+  /// data can be copied into the buffer.
+  /// </summary>
+  /// <param name="values_count"></param>
+  /// <param name="index_count"></param>
+  /// <returns></returns>
+  CooMutator MakeCooData(size_t values_count, size_t index_count);
+
+  /// <summary>
+  /// Radonly access to Csr indices
+  /// </summary>
+  class CsrView {
+   public:
+    CsrView(const Tensor& inner, const Tensor& outer) noexcept
+        : inner_(inner), outer_(outer) {}
+    const Tensor& Inner() const noexcept { return inner_; }
+    const Tensor& Outer() const noexcept { return outer_; }
+
+   private:
+    std::reference_wrapper<const Tensor> inner_;
+    std::reference_wrapper<const Tensor> outer_;
+  };
+
+  /// <summary>
+  /// Returns Csr indices readonly view
+  /// </summary>
+  /// <returns></returns>
+  CsrView AsCsr() const;
+
+  /// <summary>
+  /// This function will use Csr indices contained within the user allocated buffers.
+  /// The lifespan of the buffers must exclipse the lifespan of sparse tensor instance.
+  /// </summary>
+  /// <param name="inner_index"></param>
+  /// <param name="outer_index"></param>
+  /// <returns></returns>
+  Status UseCsrIndices(gsl::span<int64_t> inner_index, gsl::span<int64_t> outer_index);
+
+  /// <summary>
+  /// The function will allocate a single contiguous buffer and will copy values
+  /// and indices into it.
+  /// </summary>
+  /// <param name="data_transfer"></param>
+  /// <param name="data_location"></param>
+  /// <param name="values_count"></param>
+  /// <param name="values_data"></param>
+  /// <param name="inner_index"></param>
+  /// <param name="outer_index"></param>
+  /// <returns></returns>
+  Status MakeCsrData(const IDataTransfer& data_transfer,
+                     const OrtMemoryInfo& data_location,
+                     size_t values_count, void* values_data,
+                     gsl::span<const int64_t> inner_index,
```

Comment:
nit: Documentation says this will allocate a buffer and copy values and indices into it, so values_data can be const can't it?